File: web/src/lib/taxonomy.ts
/** The shared vocabulary, read from the same JSON file the pipeline uses. */
import taxonomy from "./taxonomy.json";

export type Entry = { id: string; label: string; description: string };
export type Level = { id: "learning" | "working" | "deep"; label: string; hint: string };
export type Technology = { id: string; label: string; group: string; aliases?: string[] };

export const TOPICS: Entry[] = taxonomy.topics;
export const KINDS: Entry[] = taxonomy.kinds;
export const AUDIENCES: Entry[] = taxonomy.audiences;
export const LEVELS = taxonomy.levels as Level[];
export const TECHNOLOGIES = taxonomy.technologies as Technology[];

export const TOPIC_IDS = new Set(TOPICS.map((t) => t.id));
export const KIND_IDS = new Set(KINDS.map((k) => k.id));
export const LEVEL_IDS = new Set<string>(LEVELS.map((l) => l.id));
export const TECHNOLOGY_IDS = new Set(TECHNOLOGIES.map((t) => t.id));

const LABELS = new Map(
  [...TOPICS, ...KINDS, ...AUDIENCES, ...TECHNOLOGIES].map((e) => [e.id, e.label]),
);

export function label(id: string) {
  return LABELS.get(id) ?? id;
}

/** Same folding as agent/taxonomy.technology_id: known spellings collapse to one id, and
 *  an unknown name survives as a slug so the two halves agree about what a tag means. */
const TECHNOLOGY_ALIASES = new Map<string, string>(
  TECHNOLOGIES.flatMap((t) =>
    [t.id, t.label, ...(t.aliases ?? [])].map((name) => [name.trim().toLowerCase(), t.id] as const),
  ),
);

const SLUG_OK = /^[a-z0-9][a-z0-9.+#-]{1,23}$/;

export function technologyId(name: string): string | null {
  const text = name.trim().toLowerCase().split(/\s+/).join(" ");
  if (!text) return null;
  const known = TECHNOLOGY_ALIASES.get(text);
  if (known) return known;
  const slug = text.replace(/[\s_/]+/g, "-").replace(/^-+|-+$/g, "");
  return SLUG_OK.test(slug) ? slug : null;
}

export const SOURCES = [
  { id: "hn", label: "Hacker News" },
  { id: "lobsters", label: "Lobsters" },
  { id: "blogs", label: "Engineering blogs" },
  { id: "arxiv", label: "arXiv" },
  { id: "github", label: "GitHub" },
] as const;

const SOURCE_LABELS = new Map<string, string>(SOURCES.map((s) => [s.id, s.label]));

export function sourceLabel(id: string) {
  return SOURCE_LABELS.get(id) ?? id;
}

File: agent/taxonomy.py
"""The shared vocabulary: topics, kinds, audiences, technologies, reading levels.

Loaded from web/src/lib/taxonomy.json, which the web app imports directly, so the pipeline
and the site cannot disagree about what an id means. Ids are stored in the database;
labels and descriptions are free to change.
"""
from __future__ import annotations

import json
import re
from functools import lru_cache
from pathlib import Path

# A name the model invented rather than picked: kept only if it looks like a tool's name
# and not a sentence, so it can be promoted into the list later under the same id.
_SLUG_OK = re.compile(r"^[a-z0-9][a-z0-9.+#-]{1,23}$")

PATH = Path(__file__).resolve().parents[1] / "web" / "src" / "lib" / "taxonomy.json"


@lru_cache(maxsize=1)
def _load() -> dict:
    return json.loads(PATH.read_text(encoding="utf-8"))


def topics() -> list[dict]:
    return _load()["topics"]


def kinds() -> list[dict]:
    return _load()["kinds"]


def audiences() -> list[dict]:
    return _load()["audiences"]


def technologies() -> list[dict]:
    return _load()["technologies"]


def technology_ids() -> frozenset[str]:
    return frozenset(t["id"] for t in technologies())


@lru_cache(maxsize=1)
def _technology_aliases() -> dict[str, str]:
    """Every spelling that means a known technology -> its id."""
    table: dict[str, str] = {}
    for entry in technologies():
        for name in [entry["id"], entry["label"], *entry.get("aliases", [])]:
            table[name.strip().lower()] = entry["id"]
    return table


def technology_id(name: str) -> str | None:
    """A name from the model or a reader -> a stable id, or None if it is not usable.

    Known spellings fold together (`k8s`, `Kubernetes` -> `kubernetes`). An unknown name
    survives as a slug rather than being dropped: the fixed list cannot cover every tool,
    and a slug that keeps appearing is the evidence for adding it properly.
    """
    if not isinstance(name, str):
        return None
    text = " ".join(name.split()).lower()
    if not text:
        return None
    known = _technology_aliases().get(text)
    if known:
        return known
    slug = re.sub(r"[\s_/]+", "-", text).strip("-")
    return slug if _SLUG_OK.match(slug) else None


def topic_ids() -> frozenset[str]:
    return frozenset(t["id"] for t in topics())


def kind_ids() -> frozenset[str]:
    return frozenset(k["id"] for k in kinds())


def audience_ids() -> frozenset[str]:
    return frozenset(a["id"] for a in audiences())


def level_ids() -> frozenset[str]:
    return frozenset(level["id"] for level in _load()["levels"])


def label(entry_id: str) -> str:
    """Human label for a topic, kind or technology id, falling back to the id itself."""
    for entry in topics() + kinds() + technologies():
        if entry["id"] == entry_id:
            return entry["label"]
    return entry_id
